src/java/com/puppetlabs/ssl_utils/ExtensionsUtils.java

```diff
@@ -759,42 +762,64 @@ private static AuthorityKeyIdentifier mapToAuthorityKeyIdentifier(
-                        "but was expected since 'serial_number' was provided");
-            }
-        }
-        else {
-            if (serialNumber == null) {
+        if (cert == null) {
+            if (pubKey == null && serialNumber == null) {
                 throw new IllegalArgumentException(
-                        "'serial_number' not provided for auth key identifier" +
-                        "but was expected since 'issuer' was provided");
+                        "Neither 'public_key', 'serial_number', or 'cert' provided for " +
+                        "auth key identifier.  At least one of these must be " +
+                        "provided.");
             }
 
-            GeneralNames issuerAsGeneralNames =
-                    new GeneralNames(new GeneralName(new X500Name(issuer)));
-            if (authorityKeyId != null) {
-                authorityKeyId = new AuthorityKeyIdentifier(
-                        authorityKeyId.getKeyIdentifier(),
-                        issuerAsGeneralNames,
-                        serialNumber);
+            String issuer = (String) authKeyIdMap.get ("issuer_dn");
+            if (issuer == null) {
+                if (serialNumber != null) {
+                    throw new IllegalArgumentException(
+                            "'issuer' not provided for auth key identifier " +
+                            "but was expected since 'serial_number' was provided");
+                }
             }
             else {
-                authorityKeyId = new AuthorityKeyIdentifier(
-                        issuerAsGeneralNames,
-                        serialNumber);
+                if (serialNumber == null) {
+                    throw new IllegalArgumentException(
+                            "'serial_number' not provided for auth key identifier" +
+                            "but was expected since 'issuer' was provided");
+                }
+
+                GeneralNames issuerAsGeneralNames =
+                        new GeneralNames(new GeneralName(new X500Name(issuer)));
+                if (authorityKeyId != null) {
+                    authorityKeyId = new AuthorityKeyIdentifier(
+                            authorityKeyId.getKeyIdentifier(),
+                            issuerAsGeneralNames,
+                            serialNumber);
+                }
+                else {
+                    authorityKeyId = new AuthorityKeyIdentifier(
+                            issuerAsGeneralNames,
+                            serialNumber);
+                }
             }
         }
+        else {
+            DigestCalculator digCalc =
+                new JcaDigestCalculatorProviderBuilder().build().get(
+                               new AlgorithmIdentifier(OIWObjectIdentifiers.idSHA1));
+
+            X509ExtensionUtils utils = new X509ExtensionUtils(digCalc);
+            authorityKeyId = utils.createAuthorityKeyIdentifier(new X509CertificateHolder(cert.getEncoded()));
```

Comment:
This usage of `getEncoded` here is what causes all the calling functions to need `CertificateEncodingExceptions` added to the list of possible throws. I wanted some of the below, commented functions to work, but I couldn't quite get them to work right. Comments encouraged!